satellite/gracefulexit/endpoint.go

```diff
@@ -862,3 +717,97 @@ func (endpoint *Endpoint) updatePointer(ctx context.Context, originalPointer *pb
+	exitStatus, err := endpoint.overlaydb.GetExitStatus(ctx, nodeID)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	if exitStatus.ExitFinishedAt != nil {
+		// TODO maybe we should store the reason in the DB so we know how it originally failed.
+		return endpoint.getFinishedMessage(ctx, endpoint.signer, nodeID, *exitStatus.ExitFinishedAt, exitStatus.ExitSuccess, -1)
+	}
+
+	if exitStatus.ExitInitiatedAt == nil {
+		request := &overlay.ExitStatusRequest{NodeID: nodeID, ExitInitiatedAt: time.Now().UTC()}
+		node, err := endpoint.overlaydb.UpdateExitStatus(ctx, request)
+		if err != nil {
+			return nil, Error.Wrap(err)
+		}
+		err = endpoint.db.IncrementProgress(ctx, nodeID, 0, 0, 0)
+		if err != nil {
+			return nil, Error.Wrap(err)
+		}
+
+		// graceful exit initiation metrics
+		age := time.Now().UTC().Sub(node.CreatedAt.UTC())
+		mon.FloatVal("graceful_exit_init_node_age_seconds").Observe(age.Seconds())
+		mon.IntVal("graceful_exit_init_node_audit_success_count").Observe(node.Reputation.AuditSuccessCount)
+		mon.IntVal("graceful_exit_init_node_audit_total_count").Observe(node.Reputation.AuditCount)
+		mon.IntVal("graceful_exit_init_node_piece_count").Observe(node.PieceCount)
+
+		return &pb.SatelliteMessage{Message: &pb.SatelliteMessage_NotReady{NotReady: &pb.NotReady{}}}, nil
+	}
+
+	if exitStatus.ExitLoopCompletedAt == nil {
+		return &pb.SatelliteMessage{Message: &pb.SatelliteMessage_NotReady{NotReady: &pb.NotReady{}}}, nil
+	}
+
+	return nil, nil
+}
+
+func (endpoint *Endpoint) generateExitStatusRequest(ctx context.Context, progress *Progress) (*overlay.ExitStatusRequest, error) {
+	processed := progress.PiecesFailed + progress.PiecesTransferred
+
+	if processed > 0 {
+		mon.IntVal("graceful_exit_successful_pieces_transfer_ratio").Observe(progress.PiecesTransferred / processed)
+	}
+
+	exitStatusRequest := &overlay.ExitStatusRequest{
+		NodeID:         progress.NodeID,
+		ExitFinishedAt: time.Now().UTC(),
+	}
+	// check node's exiting progress to see if it has failed passed max failure threshold
+	if processed > 0 && float64(progress.PiecesFailed)/float64(processed)*100 >= float64(endpoint.config.OverallMaxFailuresPercentage) {
+		exitStatusRequest.ExitSuccess = false
+	} else {
+		exitStatusRequest.ExitSuccess = true
+	}
+
+	if exitStatusRequest.ExitSuccess {
+		mon.Meter("graceful_exit_success").Mark(1)
+	} else {
+		mon.Meter("graceful_exit_fail_max_failures_percentage").Mark(1)
+	}
+
+	return exitStatusRequest, nil
+
+}
+
+func (endpoint *Endpoint) calculatePieceSize(ctx context.Context, pointer *pb.Pointer, incomplete *TransferQueueItem, nodePiece *pb.RemotePiece) (int64, error) {
+	nodeID := incomplete.NodeID
+
+	// calculate piece size
+	redundancy, err := eestream.NewRedundancyStrategyFromProto(pointer.GetRemote().GetRedundancy())
+	if err != nil {
+		return 0, Error.Wrap(err)
+	}
+
+	if !validateRedundancyThreshold(ctx, pointer, redundancy) {
```

Comment:
super nit - since this returns a bool, might read more smoothly as `validRedundancyThreshold`

Reply:
or it could stay as is, but maybe put the following log line and update pieces call into the function...? idk, this is very nit but I think `calculatePieceSize` could do less

Reply:
I removed the `validRedundancyThreshold` function since it's not really needed